net/strongswan/files/swanctl.init

```diff
@@ -0,0 +1,559 @@
+#!/bin/sh /etc/rc.common
+
+START=90
+STOP=10
+
+USE_PROCD=1
+PROG=/usr/lib/ipsec/starter
+
+. $IPKG_INSTROOT/lib/functions.sh
+. $IPKG_INSTROOT/lib/functions/network.sh
+
+STRONGSWAN_CONF_FILE=/etc/strongswan.conf
+STRONGSWAN_VAR_CONF_FILE=/var/ipsec/strongswan.conf
+
+SWANCTL_CONF_FILE=/etc/swanctl/swanctl.conf
+SWANCTL_VAR_CONF_FILE=/var/swanctl/swanctl.conf
+
+WAIT_FOR_INTF=0
+
+time2seconds()
+{
+	local timestring="$1"
+	local multiplier number suffix
+
+	suffix="${timestring//[0-9 ]}"
+	number="${timestring%%$suffix}"
+	[ "$number$suffix" != "$timestring" ] && return 1
+	case "$suffix" in
+	""|s)
+		multiplier=1 ;;
+	m)
+		multiplier=60 ;;
+	h)
+		multiplier=3600 ;;
+	d)
+		multiplier=86400 ;;
+	*)
+		return 1 ;;
+	esac
+	echo $(( number * multiplier ))
+}
+
+seconds2time()
+{
+	local seconds="$1"
+
+	if [ $seconds -eq 0 ]; then
+		echo "0s"
+	elif [ $((seconds % 86400)) -eq 0 ]; then
+		echo "$((seconds / 86400))d"
+	elif [ $((seconds % 3600)) -eq 0 ]; then
+		echo "$((seconds / 3600))h"
+	elif [ $((seconds % 60)) -eq 0 ]; then
+		echo "$((seconds / 60))m"
+	else
+		echo "${seconds}s"
+	fi
+}
+
+file_reset() {
+	: > "$1"
+}
+
+xappend() {
+	local file="$1"
+	shift
+
+	echo "$@" >> "$file"
+}
+
+swan_reset() {
+	file_reset "$STRONGSWAN_VAR_CONF_FILE"
+}
+
+swan_xappend() {
+	xappend "$STRONGSWAN_VAR_CONF_FILE" "$@"
+}
+
+swan_xappend0() {
+	swan_xappend "$@"
+}
+
+swan_xappend1() {
+	swan_xappend "  ""$@"
+}
+
+swan_xappend2() {
+	swan_xappend "    ""$@"
+}
+
+swan_xappend3() {
+	swan_xappend "      ""$@"
+}
+
+swan_xappend4() {
+	swan_xappend "        ""$@"
+}
+
+swanctl_reset() {
+	file_reset "$SWANCTL_VAR_CONF_FILE"
+}
+
+swanctl_xappend() {
+	xappend "$SWANCTL_VAR_CONF_FILE" "$@"
+}
+
+swanctl_xappend0() {
+	swanctl_xappend "$@"
+}
+
+swanctl_xappend1() {
+	swanctl_xappend "  ""$@"
+}
+
+swanctl_xappend2() {
+	swanctl_xappend "    ""$@"
+}
+
+swanctl_xappend3() {
+	swanctl_xappend "      ""$@"
+}
+
+swanctl_xappend4() {
+	swanctl_xappend "        ""$@"
+}
+
+warning() {
+	echo "WARNING: $@" >&2
+}
+
+add_crypto_proposal() {
+	local encryption_algorithm
+	local hash_algorithm
+	local dh_group
+	local prf_algorithm
+
+	config_get encryption_algorithm "$1" encryption_algorithm
+	config_get hash_algorithm "$1" hash_algorithm
+	config_get dh_group "$1" dh_group
+	config_get prf_algorithm "$1" prf_algorithm
+
+	[ -n "$encryption_algorithm" ] && \
+		crypto="${crypto:+${crypto},}${encryption_algorithm}${hash_algorithm:+-${hash_algorithm}}${dh_group:+-${dh_group}}${prf_algorithm:+-${prf_algorithm}}"
```

Comment:
Please reorder so it's encryption-hash-prf-dh_group
And if encryption is an AEAD algorithm, reject proposals with a hash, because that'd be an invalid proposal in IKE.